lib/_tls_wrap.js

```diff
@@ -1459,7 +1459,7 @@ exports.connect = function connect(...args) {
     tlssock._start();
 
   tlssock.on('secure', onConnectSecure);
-  tlssock.once('end', onConnectEnd);
+  tlssock.prependListener('end', onConnectEnd);
```

Comment:
This needs to run before any other end handler to properly propagate error before destroy.

There is a lot of legacy code related...

Reply:
I've never seen prependListener actually used, and am having trouble seeing it as anything other than a fragile hack. This change also doesn't seem related to the relative order of  error and close. What necessitates this?

Reply:
It's a bug in the TLS code where `close` is emitted before `error`.  A "proper" fix requires significantly more changes.

Basically it want to emit and error on a premature/unexpected `'end'`, However, the `Socket` base class has a registered `'end'` handles which will "auto destroy" the socket without the error before `onConnectEnd`  has run.